packages/inferno-router/src/Route.ts

```diff
@@ -1,123 +1,139 @@
+import { Children } from "./utils";
+import matchPath from "./matchPath";
 
-const resolvedPromise = Promise.resolve();
-
-export type IRouteHook = (props?: any, router?: any) => void;
+const isEmptyChildren = children => Children.count(children) === 0;
 
 export interface IRouteProps {
-  params?: any;
-  onEnter?: IRouteHook;
-  onLeave?: IRouteHook;
-  asyncBefore?: any;
-  path: string;
-  children: Array<Component<any, any>>;
-  component?: Component<any, any>;
-  getComponent(
-    nextState: any,
-    callback: (error: any, comp: Component<any, any>) => void
-  ): void;
+  computedMatch: any; // private, from <Switch>
+  path: any;
+  exact: any;
+  strict: any;
+  sensitive: any;
+  component: any;
+  render: any;
+  location: any;
+  children: Component<any, any>[];
 }
 
-export default class Route extends Component<IRouteProps, any> {
-  constructor(props?: IRouteProps, context?: any) {
+/**
+ * The public API for matching a single path and rendering.
+ */
+class Route extends Component<IRouteProps, any> {
+  public getChildContext() {
+    return {
+      router: {
+        ...this.context.router,
+        route: {
+          location: this.props.location || this.context.router.route.location,
+          match: this.state.match
+        }
+      }
+    };
+  }
+
+  constructor(props?: any, context?: any) {
     super(props, context);
     this.state = {
-      asyncComponent: null
+      match: this.computeMatch(props, context.router)
     };
   }
 
-  public componentWillMount() {
-    const { onEnter } = this.props;
-    const { router } = this.context;
-
-    if (onEnter) {
-      resolvedPromise.then(() => {
-        onEnter({ props: this.props, router });
-      });
+  public computeMatch(
+    { computedMatch, location, path, strict, exact, sensitive },
+    router
+  ) {
+    if (computedMatch) {
+      // <Switch> already computed the match for us
+      return computedMatch;
     }
 
-    const { getComponent } = this.props;
-    if (getComponent) {
-      resolvedPromise.then(() => {
-        getComponent({ props: this.props, router }, this._onComponentResolved);
-      });
-    }
-  }
+    invariant(
+      router,
+      "You should not use <Route> or withRouter() outside a <Router>"
+    );
 
-  private _onComponentResolved = (error, component) => {
-    this.setState({
-      asyncComponent: component
-    });
-  };
+    const { route } = router;
+    const pathname = (location || route.location).pathname;
 
-  public doAsyncBefore(params) {
-    if (this.props.asyncBefore) {
-      return this.props.asyncBefore(params);
-    } else {
-      return Promise.resolve();
-    }
+    return path
+      ? matchPath(pathname, { path, strict, exact, sensitive })
+      : route.match;
   }
 
-  public onLeave(trigger = false) {
-    const { onLeave } = this.props;
-    const { router } = this.context;
-
-    if (onLeave && trigger) {
-      onLeave({ props: this.props, router });
-    }
+  public componentWillMount() {
```

Comment:
this componentWillMount function could be behind process.env.NODE_ENV !== 'production'  to reduce production size of library